sql/wsrep_high_priority_service.cc

```diff
@@ -126,7 +126,18 @@ static int apply_events(THD*                       thd,
                         const wsrep::const_buffer& data,
                         wsrep::mutable_buffer&     err)
 {
-  int const ret= wsrep_apply_events(thd, rli, data.data(), data.size());
+  uint n_retries = 0;
+
+  int ret= wsrep_apply_events(thd, rli, data.data(), data.size());
+
+  while (ret && n_retries < wsrep_applier_retry_count) {
+    /* appling failed, retry applying events */
```

Comment:
Would not we need to roll back statement/transaction here?

Reply:
Yes, you are right. We have to roll back or undo a partially applied write set when a `wsrep_apply_events()` call fails.
The MTR test `galera.galera_retry_applying` was too simple to detect this problem.

Reply:
I have fixed this in the latest commit. 

If `wsrep_applier_retry_count` is positive, then a savepoint is created before the first apply attempt. If applying fails, the transaction is restored by rollback to savepoint before the next apply attempt.